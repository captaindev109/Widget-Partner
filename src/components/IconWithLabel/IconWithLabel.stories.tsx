import React from 'react';
import { Meta, Story } from '@storybook/react';
import styled from 'styled-components';
import { decorators } from '@/../.storybook/preview';
import Wrapper from '@/generics/Wrapper';
import IconWithLabel from '.';
import { IconWithLabelProps } from './IconWithLabel';

// A wrapper to position the component
const StoryWrapper = styled(Wrapper)``;

// Name the folder to either generics or components
const meta: Meta = {
  title: 'Components/IconWithLabel',
  component: IconWithLabel,
  decorators,
};

export default meta;

// Showcasing the component
const Template: Story<IconWithLabelProps> = (args) => (
  <StoryWrapper>
    <IconWithLabel {...args} />
    <IconWithLabel {...args} />
  </StoryWrapper>
);

export const DefaultIconWithLabel = Template.bind({});

DefaultIconWithLabel.args = {};
