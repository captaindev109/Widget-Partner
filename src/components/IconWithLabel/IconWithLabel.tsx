import { FC } from 'react';
import styled, { useTheme } from 'styled-components';
import Text from '@/generics/Text';
import Card from '@/generics/Card';
import { SVGConfigProps, TextVariantTypes } from '@/types/types';
import { SVGComponent } from '../SVGIcons/Index';

// A basic Interface or type is required for every each component
export interface IconWithLabelProps {
  iconName: string;
  position?: 'left' | 'right';
  config?: SVGConfigProps;
  label?: string;
  variant?: TextVariantTypes;
}

const StyledText = styled(Text)<{ color: string }>`
  color: ${(props) => props.color};
`;

//The main component with dynamic rendering and state
const IconWithLabel: FC<IconWithLabelProps> = ({ iconName, label, config, variant = 'Body2', position = 'left' }) => {
  const theme = useTheme();

  return (
    <Card display="flex" alignItems="center" data-testid="IconWithLabel">
      {position === 'left' && (
        <Card display="flex" alignItems="center">
          <SVGComponent iconName={iconName} config={config} />
        </Card>
      )}
      {label && (
        <Card
          display="flex"
          alignItems="center"
          paddingLeft={position === 'left' ? theme.size.padding.p16 : '0px'}
          paddingRight={position === 'right' ? theme.size.padding.p16 : '0px'}
        >
          <StyledText color={config?.color} variant={variant}>
            {label}
          </StyledText>
        </Card>
      )}
      {position === 'right' && (
        <Card display="flex" alignItems="center">
          <SVGComponent iconName={iconName} config={config} />
        </Card>
      )}
    </Card>
  );
};

export default IconWithLabel;
