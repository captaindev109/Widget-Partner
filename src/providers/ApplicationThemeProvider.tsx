import { ThemeProvider, DefaultTheme } from 'styled-components';
import GlobalStyles from '@/styles/Global';
import React from 'react';
import { ChildrenReturnType } from '@/types/types';

interface ApplicationThemeProviderProps {
  children?: ChildrenReturnType;
  theme: DefaultTheme;
}

const ApplicationThemeProvider = ({ children, theme }: ApplicationThemeProviderProps) => {
  return (
    <ThemeProvider theme={theme}>
      <GlobalStyles />
      {children}
    </ThemeProvider>
  );
};

export default ApplicationThemeProvider;
