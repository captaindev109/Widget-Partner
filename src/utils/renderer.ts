import React from 'react';
import { ConfigProps, Index, EachComponentObjectChildren, BasicAny } from '@/types/types';
import { UIComponents } from './components';

const recursivelyCreateReactComponents = (
  componentToCreate: BasicAny,
  configuration: ConfigProps,
  children: BasicAny
) => {
  if (typeof componentToCreate !== 'undefined') {
    return React.createElement(
      componentToCreate,
      configuration,
      children &&
        (typeof children === 'string'
          ? children
          : children.map((component: EachComponentObjectChildren, index: Index) => viewRenderer(component, index)))
    );
  }
  null;
};

const viewRenderer: BasicAny = (config: ConfigProps, index: Index) => {
  const children = config.children;
  const newConfig = { ...config, key: index };
  const LookForComponentInUILibrary = UIComponents[config.component];
  return recursivelyCreateReactComponents(LookForComponentInUILibrary, newConfig, children);
};

export default viewRenderer;
