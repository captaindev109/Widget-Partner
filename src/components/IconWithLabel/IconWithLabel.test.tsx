/**
 *
 *  ********** Unit & Snapshot Testing *********
 *
 * Unit test is specifiic for each functionliaty of the component
 * Snapshot test is to track Over All changes in the code since the first snap
 **/

import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import IconWithLabel from './IconWithLabel';
import defaultTheme from '@/theme/theme';
import ApplicationThemeProvider from '@/providers/ApplicationThemeProvider';

const MockComponet: React.FC = () => {
  return (
    <ApplicationThemeProvider theme={defaultTheme}>
      <IconWithLabel iconName="CheckMark" />
    </ApplicationThemeProvider>
  );
};

// describe function is a better way to organize our test cases and describe which component we are testing
describe('<IconWithLabel /> Test Cases', () => {
  // this test is a must for each component, any rendering issues with the component can be found
  test('it renders without crashing, all is well', () => {
    // render basically prints and exposes the component in nodejs environemt, it returns an object...
    render(<MockComponet />);
  });

  it('Snapshot test should pass', () => {
    const snap = renderer.create(<MockComponet />).toJSON();
    expect(snap).toMatchSnapshot();
  });
});
