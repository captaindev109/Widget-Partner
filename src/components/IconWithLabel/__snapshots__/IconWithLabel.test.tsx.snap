// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IconWithLabel /> Test Cases Snapshot test should pass 1`] = `
<div
  className="Card-sc-fb64gk-0 hskMYt"
  data-testid="IconWithLabel"
  display="flex"
>
  <div
    className="Card-sc-fb64gk-0 hskMYt"
    display="flex"
  >
    <svg
      fill="#505050"
      height="15px"
      viewBox="0 0 20 15"
      width="20px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clipRule="evenodd"
        d="M19.5486 1.51146L7.81889 14.0915L0.45874 6.41706L1.54134 5.3788L7.80354 11.9084L18.4515 0.488525L19.5486 1.51146Z"
        fill="#505050"
        fillRule="evenodd"
      />
    </svg>
  </div>
</div>
`;
